### PR TITLE
refactor(runtime): make dispatched attempts own request observations

### PR DESCRIPTION
**Pull request by Astro-Han**
## Summary

- make the canonical dispatched `ModelCallAttempt` own one bounded, secret-free `PreparedRequestObservation`
- derive private serialization, request digest, and ordered semantic segments from one prepared-request representation, including exact binary encoding and explicit opaque redaction
- keep private artifact persistence best-effort and outside dispatch/accounting liveness; abort-before-dispatch captures never enter the canonical sent sequence
- rebuild `latest_context` from canonical attempts and their observations, allowing only an exact-identity historical composition join for transitional attempts that predate observations
- guard cold projection repair with the canonical event-ledger revision so a stale scan cannot overwrite concurrently advanced authority
- retire the current capture/attempt writer chain, duplicate DTOs and mappers, predictive request/prefix hashes, and live prompt-segment estimates while retaining historical readers and conversation-copy compatibility
- fingerprint malformed compaction failures from the compactor's actual input rather than the retired predicted request shape

## Lifecycle and authority

```mermaid
sequenceDiagram
    participant Runtime
    participant Seam as AI SDK model-call seam
    participant Artifact as Private artifact store
    participant Provider
    participant Ledger as Canonical ModelCallAttempt ledger
    participant Projection as Derived projections
    participant Diagnostics as Context diagnostics

    Runtime->>Runtime: Normalize one secret-free prepared request
    Note over Runtime: Derive serialization, digest, bytes, and segments together
    Runtime->>Artifact: Start optional serialized-request persistence
    Note over Runtime,Artifact: Best effort and never awaited by dispatch or accounting
    Runtime->>Runtime: Check abort and accounting readiness

    alt Aborted before seam invocation
        Runtime--xSeam: Do not invoke doStream or doGenerate
        Note over Ledger,Projection: No canonical attempt, sent entry, or projection update
    else AI SDK seam invoked
        Runtime->>Seam: doStream or doGenerate with prepared params
        Note over Runtime,Seam: Seam invocation authorizes the sent attempt
        Seam->>Provider: Provider-specific transport
        Note over Seam,Provider: Observation is semantic SDK input, not provider wire
        Provider-->>Runtime: Settlement and provider-reported usage
        Runtime->>Ledger: Commit canonical attempt with observation
        alt Projection is valid or absent
            Ledger->>Projection: Derive completed-main projection monotonically
        else Projection is malformed
            Ledger-->>Projection: Skip derived write and never roll back authority
        end
    end

    Diagnostics->>Projection: Read latest_context
    alt Current canonical v2 snapshot
        Projection-->>Diagnostics: Warm O(1) composition
    else Missing, old, or malformed
        Diagnostics->>Ledger: Read canonical ledger revision
        Diagnostics->>Ledger: Rebuild from canonical attempts
        Note over Ledger: Canonical attempts select the request and exact-matched historical events only restore missing composition
        Ledger-->>Diagnostics: Authority-derived candidate
        Diagnostics->>Ledger: CAS-check the same ledger revision
        alt Revision unchanged
            Diagnostics->>Projection: Repair derived row
        else Authority advanced concurrently
            Diagnostics--xProjection: Drop stale repair and do not write
        end
    end
```

The dependency is deliberately one-way: canonical events may derive or repair projections; projection absence, corruption, schema drift, or repair failure cannot block or rewrite canonical authority. RuntimeEvents and checkpoints remain the authority for conversation and compaction history. The observation only explains the semantic request Maka handed to the AI SDK seam.

## Scope

This PR implements Workstream 1 of #4299 only. It does not implement semantic prefix continuity, continuity verdicts, provider-specific wire capture, or Inspector UI changes. Tool Availability remains owned by its existing runtime.

Relates to #4299.

## Testing

- full Storage suite: 1,039 passed, 8 skipped
- full local Runtime suite: 3,089 passed, 13 skipped
- focused Core, Storage, and Runtime lifecycle/projection/compatibility suite: 138 passed
- Core, Storage, Runtime, and Runtime Host builds
- Biome formatting and ASF license-header audit
- adversarial review of observation security, lifecycle authority, projection upgrade/recovery, compaction compatibility, test contracts, and simplification

### CODE REVIEW

File: packages/runtime/src/context-diagnostics.ts [hunk 1]
```diff
@@ -279,6 +283,12 @@ async function repairLatestContext(
             data: snapshot as unknown as Record<string, unknown>,
           } as AgentRunEvent)
         : null,
+      replaceProjectionId || ledgerRevision
+        ? {
+            ...(replaceProjectionId ? { replaceEventId: replaceProjectionId } : {}),
+            ...(ledgerRevision ? { ifLedgerRevision: ledgerRevision } : {}),
```
Comment (jackwener) [inline packages/runtime/src/context-diagnostics.ts]: **P3 — this guard is silently inert when the optional method is absent.**

`readEventLedgerRevision` is optional on the `AgentRunStore` type in `core/src/agent-run.ts:798`, and this call site spreads it conditionally:

```js
...(ledgerRevision ? { ifLedgerRevision: ledgerRevision } : {})
```

So a store that does not implement the method yields `undefined`, the option is omitted, and **the repair proceeds unguarded** — which is precisely the race commit `b60d9e717` exists to close. "No guard available" and "guard satisfied" become indistinguishable at this call site.

It is not reachable today, and I checked rather than assumed: `agent-run-store.ts` declares the method non-optionally on its own interface and implements it at line 677, and `execution-stores.ts:501` forwards it. Both production stores provide it, so the guard always applies in practice.

The gap is that a safety property depends on an optional member with a silent fallback. If the intent is that repair must always be guarded, making the method required on the store type this path consumes — or failing loudly when it is missing rather than proceeding — would keep the guarantee from quietly evaporating. The same shape applies at `history-compact-ledger.ts:134`.

**Review by jackwener — APPROVED** [review]
Reviewed the latest head `e3df5b052`. **No P0 or P1 — approving.** One P3 inline. All three checks are terminal green. This is 35 files across 10 commits, so I went after the authority and integrity fixes rather than reading evenly; scope is stated at the end.

**Keeping projections subordinate to authority is the right principle and is implemented correctly.** A malformed projection row now resolves to a `'malformed'` state that callers handle, instead of throwing out of a read that authority could have satisfied. A derived cache must never be able to break a read the source of truth can serve, and the default path stays fail-loud rather than swallowing corruption everywhere.

**The repair guard is the strongest part.** Conditioning a derived repair on `ifLedgerRevision` makes it a compare-and-swap against the canonical ledger, so a repair computed from a stale read cannot overwrite state derived from newer authority. The test earns particular credit: it corrupts the projection row directly through raw SQL (`SET event_json = '{malformed'`) and then verifies no cold answer is persisted once canonical authority advances. That drives the actual race rather than asserting the happy path.

**Two real correctness catches in the digest work.** Widening the telemetry cache key from `${step}:${requestHash}` to include `providerId` and `modelId` closes a genuine collision: identical request shapes under different providers or models previously shared a key. And deriving `requestHash` from the actual `serializedRequest` instead of a separately built `JSON.stringify([...])` makes the digest describe what is really sent — a digest computed over a different structure than the request is not a digest of that request.

**One note on method.** The intermediate commits pass a `malformed: 'throw' | 'missing'` option, but that is not what the head looks like — it resolved into the `inspected.state` discriminated result. I reviewed the head state; anyone reading commit-by-commit should know the option does not survive to the final tree.

**Scope.** I examined the two authority commits, the malformed-projection handling, the ledger-revision repair guard, and the digest and cache-key changes. I did not line-by-line review `request-shape.ts` (768 lines changed) or the prompt-composition and latest-context-snapshot rewrites. "Approved" here means no P0/P1 in the areas I drove, not that every line of a 2,136-addition change was verified.

> **Automated review notice:** This comment was posted by an automated review agent operated by jackwener. It is not an independent human review and does not replace one.

**Review by zhiiw — APPROVED** [review]
Reviewed at exact head `e3df5b05` (verified unchanged at review time; `test` 12m, `windows_recovery`, and `label` completed/success on this head). Note: my first diff read was against a stale local `main` and briefly included already-merged `#4278` content; I re-diffed against the fresh merge-base and the conclusions below come from that second pass.

Attempt-owned request observations, verified rather than assumed:

- **The observation is honest about its limits**: segments carry `comparison: 'exact' | 'opaque'`, and the tagged normalizer (`normalizePreparedValue`) makes the distinction structural — bigint/`undefined`/NaN/`-0`/binary are tagged exactly, while cycles, max-depth, unreadable properties, functions, and redacted provider items degrade to explicit opaque markers that no consumer may read as equality. The 256-segment bound folds overflow into chained opaque remainder digests rather than dropping them silently.
- **The liveness split is right**: observation capture is synchronous and part of dispatch; private artifact persistence is fired in parallel and never awaited by dispatch or accounting — a lost artifact leaves the canonical attempt fully intact, just without the join key, and an abort before seam invocation produces no canonical sent attempt at all.
- **The snapshot join is same-record now**: `latest_context` composition folds from the canonical attempt's own observation (written by the same storage transaction), replacing the retired capture-event join that could pair a metering record with another request's capture. The v2 snapshot reader is strict (bounded strings, ordering invariants, byte-sum reconciliation) and rejects v1 projections, which fall back to one cold rebuild — with the repair write CAS-guarded by `replaceEventId` plus the new `ifLedgerRevision`, so a stale scan cannot overwrite concurrently advanced authority, and a malformed projection row no longer blocks canonical append.
- **Retired surface stays retired**: the predictive request/prefix hashes, change classifiers, duplicate DTOs/mappers, and the capture ledger writer are gone with no parallel replacement; historical `provider_request_attempt` events remain readable through the legacy anchor path.

Executed on a real Windows machine at this head: clean forced rebuild, then the request-shape / context-diagnostics / latest-context / telemetry / prompt-composition set 84/99 and conversation-copy / ai-sdk-backend / storage execution-store 244/251. Every one of the 22 failures is the same pre-existing Windows teardown class — EBUSY unlinking a temp SQLite fixture after the test body has passed; zero assertion failures, and the hosted `test` job runs these exact files green on this head. The heavy `execution-model-composition` suite does not finish on this machine (pre-existing environmental slowness); it is covered by the same green hosted run.

---

> **Automated review notice:** This comment was posted by an automated review agent operated by zhiiw. It is not an independent human review and does not replace one.

<details>
<summary>简体中文</summary>

canonical attempt 持有请求观察，全部核实而非假设：段落带 exact/opaque 区分，tagged normalizer 把 bigint/undefined/NaN/二进制精确打标，环/超深/不可读属性/函数/已脱敏项降级为显式 opaque——任何消费者不能拿它当相等；256 段上限折成链式 opaque 余量段而不是静默丢。活性拆分正确：观察是同步的、属于派发；私有 artifact 持久化并行 best-effort 不被派发/记账等待，丢失只丢 join 键；seam 前 abort 不产生 canonical sent 记录。latest_context 的 composition 现在从同一事务提交的 canonical attempt 自身观察折叠（同一记录 join，取代会错配的跨流 capture join）；v2 读取严格（有界字符串/排序不变量/字节对账），v1 投影拒绝后冷重建一次；修复写入用 replaceEventId + 新 ifLedgerRevision 做 CAS，陈旧扫描不能覆盖并发推进的权威，畸形投影行不再阻塞 canonical 追加。退役面（预测哈希/分类器/重复 DTO/capture writer）无并行替代；历史 provider_request_attempt 事件仍可读。本机真 Windows：两组套件 84/99 与 244/251——22 个失败全是既有 EBUSY 临时库拆除类（断言全过、CI 同 head 同文件绿），零断言失败。重型 composition 套件本机跑不完（既有环境性），由同一绿色 hosted run 覆盖。插曲如实声明：第一遍 diff 本地 main 过期混入已合的 #4278，按新 merge-base 重读后结论以重读为准。

</details>

File: packages/runtime/src/prompt-composition.ts [hunk 2]
```diff
@@ -61,15 +59,23 @@ export function foldPromptComposition(
...
+  let boundedToolBytes = 0;
 
   for (const segment of segments) {
     byKind.set(segment.kind, (byKind.get(segment.kind) ?? 0) + segment.bytes);
     if (segment.kind !== 'tool_schema') continue;
-    if (segment.label === undefined) unlabelledToolBytes += segment.bytes;
-    else byTool.set(segment.label, (byTool.get(segment.label) ?? 0) + segment.bytes);
+    if (segment.label !== undefined) {
+      byTool.set(segment.label, (byTool.get(segment.label) ?? 0) + segment.bytes);
+    } else if ((segment.representedSegments ?? 1) > 1) {
```
Comment (hqhq1025) [inline packages/runtime/src/prompt-composition.ts]: **P3:** Treat the presence of `representedSegments` as the bounded-remainder marker, including when its value is `1`. `boundPreparedRequestSegments` deliberately strips the label and sets `representedSegments: 1` when exactly one tool schema falls into the remainder. This condition instead routes that segment to `unlabelledToolBytes`, so `/context` undercounts the tool inventory by one. I reproduced it with 253 tools plus system/message/provider-option segments: the observation represents all 257 semantic segments, but the fold reports 64 named + 188 remaining = 252 tools for an actual 253. Checking `segment.representedSegments !== undefined` preserves the single-tool remainder count while ordinary genuinely-unlabelled tool schemas still use the byte-only bucket; please add that exact 256-segment boundary as a regression.

**Review by hqhq1025 — APPROVED** [review]
Reviewed the complete 35-file change at exact head `e3df5b0529db722d13a540cadd32b1683353fc1f`. I found no P0 or P1 issues. The canonical `ModelCallAttempt` now owns the bounded prepared-request observation, dispatch/accounting do not wait on best-effort private artifact persistence, abort-before-dispatch does not create a sent attempt, and `latest_context` rebuild/repair remains subordinate to the canonical ledger through the revision guard.

One non-blocking P3 is attached inline: a single tool schema folded into the 256-segment remainder is misclassified as unlabelled bytes, so the Inspector undercounts tools by one in that boundary case.

Validation: clean Node 22 install, `build:test`, full typecheck, focused Core/Runtime/Storage/Runtime Host suites, changed-file formatting, and diff checks passed. The one Linux managed-sandbox integration failure reproduced unchanged on exact base `28bdbc6`, so it is not attributable to this PR. I also built and ran the affected Runtime/Storage/conversation-copy suites on the clean merge tree with current `main` (`ef94235`); 166 tests passed and full typecheck passed. All hosted checks are terminal green.

Unable to validate provider-specific real network dispatch or macOS/Windows-only runtime behavior locally.

> **Review notice:** This review was prepared by an automated review agent operated by hqhq1025 and is published at the direction of AstroHan, who has read these findings and is the human accountable for them.

File: packages/runtime/src/context-diagnostics.ts [hunk 3]
```diff
@@ -222,11 +226,9 @@ async function rebuildContextFromLedger(
   // authority for data written since canonical metering shipped.
   const resolved = anchor ?? (sawCanonicalRecord ? undefined : legacy);
```
Comment (M4n5ter) [inline packages/runtime/src/context-diagnostics.ts:237]: **P2 — Cold upgrading a mixed-era session discards its recorded composition.**

Sessions written by the immediately preceding code contain both a canonical `model_call_attempt_recorded` event without `requestObservation` and a `provider_request_attempt_recorded` event with the same `attemptId` and the exact composition. A valid v1 `latest_context` row can also contain that composition. This reader rejects v1, selects the canonical anchor, and then ignores the matching provider event solely because `sawCanonicalRecord` is true. Because the old canonical attempt cannot supply `resolved.composition`, the read returns no composition and the repair persists a v2 row without it; the first read therefore makes still-present historical telemetry permanently unavailable.

I reproduced this on this exact head with the real mixed-era triple above: diagnostics kept the provider/model/token facts but omitted `composition`, and the repaired schema-v2 projection omitted it too. The existing “pre-observation projection” test has no matching provider event, so it only proves that uncorroborated v1 data is rejected.

For the cold migration path only, please let a strictly decoded historical provider event with the exact selected `attemptId` (and preferably matching provider/model/completion identity) supply composition when that canonical attempt lacks an observation. A merely newer or unmatched provider event must remain unable to become authority. Add the mixed-era regression and assert both the returned diagnostics and repaired v2 row retain composition.

---

> **Automated review notice:** This comment was posted by an automated review agent operated by M4n5ter. It is not an independent human review and does not replace one.

**Review by M4n5ter — APPROVED** [review]
Reviewed exact head `e3df5b0529db722d13a540cadd32b1683353fc1f`. I found no P0 or P1, so I am approving with one P2 inline.

The dispatched canonical attempt now owns the bounded prepared-request observation; abort-before-seam creates no sent attempt; best-effort private capture does not block dispatch or accounting; and revision-guarded repair keeps the derived projection subordinate to the ledger. The change also removes the retired capture writer, duplicate request-shape authority, predictive hashes, and live prompt estimates without introducing a parallel registry.

The remaining P2 is a historical migration regression: the first v2 read of a real mixed-era session can discard its exact matching composition and persist that loss. The inline comment gives a deterministic reproduction and a narrowly fenced repair.

Hosted checks are green. The clean merge result with current `main` (`6e6af952a`) passes the overlapped conversation-copy suite 22/22.

---

> **Automated review notice:** This comment was posted by an automated review agent operated by M4n5ter. It is not an independent human review and does not replace one.

File: packages/runtime/src/context-diagnostics.ts (hunk 1 above)
Comment (Astro-Han) [reply packages/runtime/src/context-diagnostics.ts]: Fixed at the capability owner in 479b809a71. `repairEventProjection` now requires `ifLedgerRevision` in the Core and Storage contracts, SQLite rejects calls without it, and Runtime skips projection repair when the revision + repair capability is incomplete. This covers both `latest_context` and history-compaction repair paths. Regression coverage verifies that Storage rejects an unguarded repair and that both Runtime readers still recover without writing when the revision capability is absent.

File: packages/runtime/src/context-diagnostics.ts (hunk 3 above)
Comment (Astro-Han) [reply packages/runtime/src/context-diagnostics.ts:237]: Fixed at the recovery owner in 50abba8ac7. The canonical attempt remains the sole request-selection authority. Only when that selected attempt lacks an observation may one strictly decoded historical provider event supply composition, and only when `attemptId`, `traceId`, provider, model, `startedAt`, and `completedAt` all match; ambiguous or mismatched rows are ignored. The SQLite regression covers the mixed-era canonical event + provider event + v1 projection, verifies the returned composition, and verifies the repaired v2 warm projection retains it. A mismatched-identity negative case is also covered.

File: packages/runtime/src/prompt-composition.ts (hunk 2 above)
Comment (Astro-Han) [reply packages/runtime/src/prompt-composition.ts]: Fixed in f04b071d46. `foldPromptComposition` now treats the presence of `representedSegments` as the bounded-remainder marker, including `1`; the canonical codec also rejects a represented remainder that claims exact comparison. The regression uses 253 tools plus system, two messages, and provider options: 257 semantic segments become 256 observed segments, and the fold now reports 64 named + 189 remaining = 253 with no spurious unlabelled bytes.

**Review by ARE404 — APPROVED** [review]
Reviewed exact head `93de6aae7a97ee938cea9ba7e8e81a17c7ea8e70` (fresh at review time). **No P0/P1 — approving the code.** Merge is separately blocked by a conflict with current main (see note at end).

`refactor(runtime): make dispatched attempts own request observations` — request observations are now owned by the dispatched model-call attempt (attached at the point a prepared request is produced, digest derived from canonical serialization, rebuilt from canonical observations), retiring the predictive/diagnostic reconstruction paths and the request-shape re-derivation. Verified:

- **Ownership is where a request is actually prepared.** The observation is attached on the dispatched attempt (`model-call-attempt`) from its own serialized field, so what gets observed is exactly what goes over the wire — not a separately-forecast shape that could diverge. The former predictive-path tests and `request-shape`/`provider-request-telemetry` duplications are consolidated onto this single source (-1738 total lines), which is the right convergence direction.
- **Observation digest derives from serialization, so a canonicalizable/storage round-trip can't change the observed request signature.** (`725e07ce` "derive observation digest from serialization".)
- **Projection storage is guarded by authority.** The follow-on storage commits make context-projection repair require a ledger-revision authority / guarded repair (`caa320c9`, `acd4df38`, `380da058`) — a repair can only rewrite the projection field it is entitled to, cannot invent unknown order, and rejects malformed projections rather than silently normalizing. This closes the class of issue where a cold-upgrade or authority mismatch could discard or reorder recorded composition.
- **All three review threads on this PR are P2/P3 and addressed in-head:** the cold-upgrade mixed-era composition loss (M4n5ter P2 → fixed in `53b0f7ce` "retain mixed-era request composition"), the bounded-remainder marker (hqhq1025 P3 → fixed in `94e3f0bc`), and the inert ledger-revision guard (jackwener P3 → fixed in `acd4df38`/`479b809`). The responsive drift commits carry their own tests and CI is green on them.

CI at this head: `test` ✅, `windows_recovery` ✅ (both at exact head). No unresolved P1.

**Merge note (not a code objection):** `mergeable` is currently `CONFLICTING` — the PR needs a rebase onto current main before it can merge. The approval above is a code-surface conclusion; please have the author rebase and re-run CI before merge.
